TextToolbar Control for Editing RichEditBox

WilliamABradley opened the pull request:
- TextToolbar w/ MarkDown & RichText Support from #1107.
- Shifted from Old branch where I had made a mess due to overwriting my other PR, you can find the original PR and commits at #1109. 
- Currently has a Sample Page and a basic RichTextFormatter Class that could be populated further by whoever is willing to help make this a Universal TextToolbar.

[![Editor](https://s20.postimg.org/ou9ksxyod/Editor.png)](https://postimg.org/image/sqmwoxjnt/)

dnfclas commented:
@WilliamABradley,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

skendrot commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/TextToolbar/TextToolbarPage.xaml.cs:
```diff
@@ -0,0 +1,96 @@
+        }
+
+        private void EditZone_TextChanged(object sender, RoutedEventArgs e)
+        {
```
No need to change the preview text every time the text changes, only need to change when the pivot changes

WilliamABradley replied:
@skendrot  fair enough

skendrot commented:
Things to remember with formatting is it's more than just formatting selection. If I hit the bold button or Ctrl-B it should make all of the new text I type bold until I turn it off. Is this handled?

skendrot commented on Microsoft.Toolkit.Uwp.UI.Controls/TextToolbar/Formats/Format.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace Microsoft.Toolkit.Uwp.UI.Controls.TextToolbarFormats
+{
+    public enum Format
```
I don't think tying the control to these types makes sense. I would rather see it fire events for the buttons like BoldActivated/BoldDeactivated or FormatActivated event with a Type parameter of "Bold". Then controls could be updated to support the toolbar.  Of course we cannot change the RichEditBox, so maybe it can support that out-of-the box

WilliamABradley replied:
@skendrot the Reason for the Formatter class, is that different formats require different sets of buttons. E.g. RichText supports Underline, while Markdown doesn't. Therefore, RichText has an underline button and Markdown doesn't. This also allows users to define their own Formatter class with their own styling and Default Button map.

Perhaps the abstract methods can be removed, and attached to button directly, when someone adds a CommonButton to their ButtonMap?

skendrot replied:
Understood. All the more reason **not** to tie the toolbar to any particular controls and instead allow for buttons to removed and/or added and fire events when they are activated

WilliamABradley replied:
That would mean that the user would have to add their own code to achieve the same functionality, which would defeat the point of having a control that is universal as per discussion at #1107. If not using a formatter class, how do you propose supporting both Rich Text and Markdown? With what you are saying, the user might as well just use a CommandBar and create their own from scratch.

skendrot replied:
There isn't a way to create one control to be "universal" for any text entry. What happens if I want this toolbar for a regular Textbox and not a RichTextbox? TextBox still supports bold, italics, hyperlinks, colors, etc. 

Having code paths like 
```
if(Format == RichText)
{
    DoRichText();
}
else if(Format == Markdown)
{
    DoMarkdown();
}
else if(Format == Textbox)
{
    DoTextBox()
}
```
Is a perfect example of violating the Open/Closed principle of software development. 

If I'm the only person who thinks it should be "more generic" then throw my suggestion away

WilliamABradley replied:
If you actually look at the formatter implementation, you will see that my code isn't like that. If a user wants to implement their own experience, they can either inherit the Formatter class, defining their own Default buttons, how to deal with Bold, etc. They can also insert additional custom buttons via XAML or code behind, remove default buttons via XAML or code behind.

The only purpose the Format enum has, is to Instantiate the RichText Formatter or the Markdown formatter. Set it to Custom, and then you can provide it your own inherited Formatter.

I have ensured that are plenty of places for the user to be able to adapt the code to suit their needs.

It is generic, yet you can still achieve similar functionality.

WilliamABradley commented:
@skendrot This was just the initial functionality I was setting up, considering things such as Accelerator keys, after the initial code was setup.

Again, accelerator keys would probably be of better use in the Formatter class, so that Different default Buttons can receive their own key. Such as the Code Button, which wouldn't exist in the Rich Text version of the Toolbar, could get CTRL + "[", or something like that.

WilliamABradley commented:
@skendrot I have implemented Rudimentary Shortcut keys, however, I need to use reflection to disable the RichEditBox's internal KeyDown interactions, so that Markdown can use combinations such as CTRL + R & CTRL + L, as it doesn't support Right to Left Text, Or I could just stray away from using those keys.

I still think having a Formatter class to format the text, as RichEditBox, doesn't have the support, or the code split could be moved, so there is two Controls, an Inherited RichEditBox, that uses a Formatter Class, and then the TextToolbar will just trigger the methods in this Inherited RichEditBox. Which is a more natural location for a Formatter Style class, however, this is a simpler implementation.

nmetulev commented:
Love the control. I'm slowly reviewing the PR. In the meantime, I keep running into a marshaling exception that you should probably take a look. It happens almost every time I play with the markdown format, switch to rich format and click on to view the xaml code, (or go back between xaml code and properties)

```
System.Exception occurred
  HResult=0x8001010E
  Message=The application called an interface that was marshalled for a different thread. (Exception from HRESULT: 0x8001010E (RPC_E_WRONG_THREAD))
  Source=<Cannot evaluate the exception source>
  StackTrace:
   at Windows.UI.Xaml.DependencyObject.ClearValue(DependencyProperty dp)
   at Microsoft.Toolkit.Uwp.UI.Helpers.DependencyPropertyWatcher`1.Dispose() in C:\Users\nikola\source\UWPCommunityToolkit\Microsoft.Toolkit.Uwp.UI\Helpers\DependencyPropertyWatcher.cs:line 79
   at Microsoft.Toolkit.Uwp.UI.Controls.TextToolbarSymbols.Symbol.Finalize() in C:\Users\nikola\source\UWPCommunityToolkit\Microsoft.Toolkit.Uwp.UI.Controls\TextToolbar\Symbols\Symbol.cs:line 56
```

nmetulev commented:
We should not be changing the behavior of REB, so CTRL+R/L, should continue to support right to left. Strike-through does not usually have a keyboard shortcut, so we can leave it without, or assign some other key.

It seems like you've implemented quite a bit of the rich text formatter, why not finish it ;)

nmetulev commented:
Please take a look at keyboard/controller navigation as well. Tab is working, but it looks like separators are also capturing focus. The controller doesn't work at all, the entire toolbar captures focus instead.

WilliamABradley commented:
@nmetulev Whoops, I had only just added the DependencyWatcher in order to grey out the Custom List Icons. I'll fix that right away, I'll patch the separators capturing focus too. As for Controller Navigation, I can't imagine in it's current state that this control to be very helpful anyway. 

I would finish Implementing RTF, however, I just don't have time. I have 5 days to Uni starts again, and I want to finish updating my app, TVPLAN. :)

WilliamABradley commented:
@nmetulev All issues fixed in my latest Commit. I even through in Automation.Name for the Toolbar Buttons.

skendrot commented:
WilliamABradley I'm not too worried about shortcuts keys, more that the toolbar handles more than changing selection. When a control uses this toolbar, and a tool is activated like Bold, then any new text entered from the location that bold was activated becomes bold.

WilliamABradley commented:
@skendrot That experience is achieved when using Rich Text format on the TextToolbar, however, due to the design of Markdown formatting, it is not as feasible.

skendrot commented:
So adding two *'s before and after the caret position isn't feasible?

WilliamABradley commented:
@skendrot no it is, but it will damage the ability to easily reverse it by pressing the button again. It is current designed mimicking Github's toolbar, test that implementation against mine.

WilliamABradley commented:
What is going on with this PR currently???

nmetulev commented:
@WilliamABradley, most of the core team is involved with the //Build conference happening next week and it's one of the reason why we have slowed down here. I expect to return back to reviewing this PR post Build.

WilliamABradley commented:
@nmetulev Awesome, thanks anyway.

WilliamABradley commented:
@nmetulev Is it worthy of a review now?

nmetulev commented:
It's on my list, planing to go through it in the next week or so.

WilliamABradley commented:
@nmetulev Sweet

nmetulev commented:
I've been slowly going though this PR, and also attempting to implement the RichText format to make sure I understand your architecture. I'm still going through everything, but wanted to share my initial thoughts so you can have at least some feedback. Overall, It's a great control, and I like the expandable nature of it all. 

For markdown:
* when clicking on unordered or ordered list and attempting to delete the list by using backspace it won't work. For ordered list, it keeps increasing the order the more you backspace
* creating a new link  while text is selected should  place the selected text in the label box
* Creating a header when the cursor is not at the beginning should add the # at the beginning of the line
*  When quote is toggled, it should automatically un-toggle when the cursor is moved from that line

For creating a new format, I want to be able to tell when the cursor position changes and toggle/un-toggle buttons depending on the cursor location and the format there. There seems to be a lot of things happening for simply toggling a ToolbarButton, that needs to be simplified. IsToggled, IsToggleable, Toggled - it's difficult to understand how they all work together, are they all necessary.  I can still toggle when IsToggleable is false. Toggling them manually stamps over other functionality, such as activate, and the behavior is almost unpredictable

That's it for now :)

WilliamABradley commented:
@nmetulev sounds like worthy things to fix, I'll see if I can get around to them by next week, I have two assignments by Friday, so I'm a bit tied up currently, but I'll fix them ASAP.

nmetulev commented:
Up?

nmetulev commented:
Doesn't look like this will make it for 1.5, so moving to 2.0. Let me know otherwise.

WilliamABradley commented:
Ok, I have just finished my last exam, so I will be able to resume work on this again. I think it might have to wait until 2.0 anyway.

nmetulev commented:
Feel free to continue working on it even now. We just won't merge it until after 1.5 ships

WilliamABradley commented:
@nmetulev I have added the changes you have suggested.
All except for the toggling off elements such as List, Ol, and Quote, when moving the mouse cursor away.
This is because the code required for this is very complicated and hard to implement.

However, You can now simply backspace the new line, and it will Toggle the buttons off.

nmetulev commented:
Thanks @WilliamABradley, I'll take a look this week

WilliamABradley commented:
@nmetulev Any News?

nmetulev commented:
@WilliamABradley, apologies, I thought I had responded. This is what I found in my testing:

For markdown:
* When trying to bold a selected word that ends with space (such as when you double tap a word to select), it should be smart enough to know to not include the space
* Adding a link, it splits in two lines which doesn't render
* Code button only adds one `, should add three, otherwise it will not render
* When list (ordered and un-ordered) is selected and cursor is moved, the button remains active. Likewise, when selecting ordered and then un-ordered (and vise-versa) both stay active but only un-ordered works

RichText seems to be completely broken, doesn't seem to work at all and throws exceptions.

WilliamABradley commented:
@nmetulev I could get working for most of those, However, determining if the cursor is in a list for Markdown is very complicated logic, so it might be easier to just switch it to GitHub logic, where it won't auto iterate, but just set the line beginnings of selected text.

nmetulev commented:
I'm ok with the way GitHub has done it, so that sounds good to me. Another option is to just disable the ordering once the cursor has moved once and not worry about detecting if the new position is a list.

WilliamABradley commented:
@nmetulev I have fixed the issues you have discovered.
The Listing function for Markdown is now much simpler.

With the Code Button, if on one line, it will add \`   \`  around selected text.
If it detects multiple lines, it will add the multi-line \`\`\` quotes around the whole group. (This may have been broken by the Toggling functions, that have now been removed.

Hopefully we can get this in before [v2.0](https://github.com/Microsoft/UWPCommunityToolkit/milestone/8)

I fixed the issue with Rich Text, turns out that I forgot to Instantiate the Rich Text Default Button Actions.

Link Creation and Markdown Text Selection formatting have also been fixed.

WilliamABradley commented:
@skendrot Will you review this?

nmetulev commented:
Thanks @WilliamABradley, will review. This is a large PR and will take some time

Quick review: 
On markdown: pressing bold/italics, etc while nothing is selected causes exception and app crashes. 

On richedit: 
* underline button stays active no matter what once pressed
* link throws exception 
* not sure what WIP does

WilliamABradley commented:
@nmetulev I have fixed the bug you encountered with Markdown.

I thought you were implementing Rich Text?

The WIP was placed there to let you know that Rich Text Editing isn't fully implemented.
We could disable the Rich Text Formatter for the time being, if you don't want to implement it.

I have fixed Underline for Rich Text however.

WilliamABradley commented:
@nmetulev I threw in the other basic functions of Rich Text Formatting. All Current buttons are working mostly (Apart from links becoming links).

It works as it is, it just doesn't support other functions, such as Font Size, Font Family, etc.

michael-hawker commented:
For Strikethrough, can we use Ctrl+- (minus) like OneNote does?

WilliamABradley commented:
@michael-hawker Done :)

WilliamABradley commented on Microsoft.Toolkit.Uwp.UI.Controls/TextToolbar/TextToolbar.Events.cs:
```diff
@@ -0,0 +1,318 @@
+                {
+                    if (e.Handled)
+                    {
+                        Editor.Document.Undo();
```
This is to handle overrides of default shortcuts, such as CTRL + - for strikethrough. Since this shortcut appears to delete text by default, this is used to undo the delete and apply the strikethrough if there is a ToolbarButton that handles it, this is improved in me next commit.

WilliamABradley commented on Microsoft.Toolkit.Uwp.UI.Controls/TextToolbar/Formats/RichText/RichTextFormatter.cs:
```diff
@@ -0,0 +1,170 @@
+            ButtonActions = new RichTextButtonActions(this);
+        }
+
+        public void FormatUnderline(ToolbarButton button)
```
This is because Underline is not a Common Button, I have moved it to ButtonActions, it just complicates how to assign that method to the Button.

WilliamABradley commented:
@nmetulev I have applied the changes you have requested.
I didn't make a logo, because I assumed that they were created by a designer on the Toolkit Team (Since they are all very Uniform). Are there style guidelines? Or Could someone help create one?

WilliamABradley commented on Microsoft.Toolkit.Uwp.UI.Controls/TextToolbar/TextToolbar.Events.cs:
```diff
@@ -0,0 +1,318 @@
+                        Editor.Document.Redo();
+                    }
+
+                    e.Handled = args.Handled;
```
This has been fixed.

WilliamABradley commented on Microsoft.Toolkit.Uwp.UI.Controls/TextToolbar/TextToolbar.Events.cs:
```diff
@@ -0,0 +1,318 @@
+        /// </summary>
+        /// <param name="obj">TextToolbar</param>
+        /// <param name="args">Property Changed Args</param>
+        public static void OnFormatTypeChanged(DependencyObject obj, DependencyPropertyChangedEventArgs args)
```
Fixed.

WilliamABradley commented on Microsoft.Toolkit.Uwp.UI.Controls/TextToolbar/ToolbarItems/ToolbarButton.cs:
```diff
@@ -0,0 +1,225 @@
+    /// <summary>
+    /// An Interactive button in the TextToolbar, to perform a formatting task.
+    /// </summary>
+    public sealed class ToolbarButton : AppBarButton, IToolbarItem, INotifyPropertyChanged
```
Woops

WilliamABradley commented:
@Odonno If you have free time, this could do with a review :)

WilliamABradley commented on Microsoft.Toolkit.Uwp.UI.Controls/TextToolbar/TextToolbar.Events.cs:
```diff
@@ -0,0 +1,317 @@
+                    }
+
+                    ShortcutRequested?.Invoke(this, args);
+                    e.Handled = args.Handled;
```
What it does is it finds any button that has a key that matches the CTRL shortcut, if it overrides an Editor Keydown, then it undoes the last action (as the keydown event might have changed something), it then checks if there is a selection after the undo, if there isn't, then nothing has changed, so redo the last action. This is just for overriding CTRL + - at the moment, as it seems to erase the last word by default. 
If a CTRL key wasn't overridden, it won't do anything.

nmetulev replied:
That makes sense. For some reason I read this line as `e.Handled = true`. :)

WilliamABradley commented:
@nmetulev I think I might have fixed the issue, when creating NuGets, my NuGet doesn't seem to like GitVersion's naming, so they don't show up and I can't test them.

I also changed the KeyDown Handled Property, so it only becomes true (and gets modified) if there has been a matching button that has been activated. This will prevent it from becoming e.Handled = false, when overriding and already handled.

nmetulev commented:
Thanks, I'll try it out. BTW, you should be able to grab the nugets from appveyor as well, in the artifacts section. [Here are the ones from the last commit.](
https://ci.appveyor.com/project/dotnetfoundation/uwpcommunitytoolkit/build/1.5.1-PullRequest01116.530.build.1486/artifacts)

WilliamABradley commented:
I feel silly now, Its because they had a pre-release flag on them. :(

WilliamABradley commented:
Ok, it works without a specified Format now, but I seem to have lost Design Time viewing.

WilliamABradley commented:
@nmetulev Everything is working except for the Designer, I'm a little stumped with this one.
I'm getting `Delegate to an instance method cannot have a null this`, when trying the Control with Project Code on.

Do you see any potential issues?
I have to go to Uni now, but I can try more fixes later.

WilliamABradley commented:
@nmetulev I fixed Design Time, it should all be working now :)

Currently Hyperlinking with Rich Text, ApplyDisplayUpdates don't seem to update the UI, if you try to Open the Link creator on text that has already created a Link, but it isn't showing, it will turn blue in the Link Creator.

One fix for this would be to Copy the RTF to a string, and then restore it (This might take time on a big document).
Otherwise, I'm not sure what is stopping it from making the link blue.

nmetulev commented:
I noticed that about the link, the link still seems to work though. It seems like a Document bug. I'd recommend creating an issue for us to track once this PR is merged.

I have another request for you. I was playing with the API and the RemoveDefaultButtons property doesn't seem natural, kinda counterintuitive to add in order to remove. Wondering if there is another way? Maybe a property on the button. Kinda like `Toolbar.DefaultButtons.GetButton(type)?.IsVisible = false;` Just an idea. Thoughts?

WilliamABradley commented:
Yea that makes more sense, the sole purpose, was that it allowed you to remove buttons in XAML. The same effect can probably be achieved this way.

WilliamABradley commented:
@nmetulev I have made the changes,
Now you can get a Default Button from the Toolbar by using `Toolbar.GetDefaultButton(ButtonType)`, which returns the ToolbarButton Instance (Based on AppBarButton). So you can set Visibility to Collapsed.

nmetulev commented:
That works much better :)

What is the purpose of ButtonModifications?

WilliamABradley commented:
@nmetulev It allows you to do the same thing from XAML.
```xaml
<controls:TextToolbar x:Name="Toolbar" Editor="{x:Bind Editor}">
    <controls:TextToolbar.ButtonModifications>
        <buttons:DefaultButton Type="Headers" IsVisible="False"/>
    </controls:TextToolbar.ButtonModifications>
```
and
```c#
var button = Toolbar.GetDefaultButton(ButtonType.Headers);
button.Visibility = Windows.UI.Xaml.Visibility.Collapsed;
```
Are Equivalent.

WilliamABradley commented:
So now we need a second review right?

nmetulev commented:
Yeah, we need two approvals before we merge

WilliamABradley commented:
I'm not sure if @skendrot is available to do this, is there someone else?

shenchauhan commented:
I'll start reviewing this :)

shenchauhan commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/TextToolbar/IsFormat.cs:
```diff
@@ -0,0 +1,33 @@
+    {
+        public object Convert(object value, Type targetType, object parameter, string language)
+        {
+            return (Format)value == Format;
```
You could get an invalid cast exception here. Add some defense against that.

WilliamABradley replied:
This was a quick converter in the Sample App, but I could make it available for use with the toolbar.

shenchauhan commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/TextToolbar/IsFormat.cs:
```diff
@@ -0,0 +1,33 @@
+
+        public object ConvertBack(object value, Type targetType, object parameter, string language)
+        {
+            throw new NotImplementedException();
```
Is there a chance this code could ever get called? Would people want to write back to a backing property? If so I'd add the convert back code too.

WilliamABradley replied:
You can't exactly infer the type of Format from a Boolean, it Could RichText, Markdown or Custom

shenchauhan commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/TextToolbar/TextToolbarCode.bind:
```diff
@@ -0,0 +1,54 @@
+        var linkOpen = Task.Run(() => Launcher.LaunchUriAsync(new Uri(e.Link)));
+    }
+    catch
+    {
```
I'm pretty sure this will not be reachable. Check out : https://msdn.microsoft.com/en-us/magazine/jj991977.aspx

WilliamABradley replied:
The link works, do you mean the catch if the link isn't a valid Uri?

shenchauhan replied:
yup, as it was running in the Task.

shenchauhan commented on Microsoft.Toolkit.Uwp.UI.Controls/TextToolbar/Formats/RichText/RichTextButtonActions.cs:
```diff
@@ -0,0 +1,147 @@
+        public void FormatUnderline(ToolbarButton button)
+        {
+            var format = Formatter.SelectionFormat;
+            if (!button.IsToggled)
```
A lot of repetitive logic. Consider a high order function.- not a must though, just makes reading easier.

WilliamABradley replied:
Some of the properties on these a very slightly different though, such as Underline and list types.

shenchauhan commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/TextToolbar/TextToolbarPage.xaml.cs:
```diff
@@ -0,0 +1,107 @@
+                }
+            };
+
+            Toolbar.CustomButtons.Add(demoButton);
```
I found a bug:
- Switch to dark
- IsEnabled = Off
- Format: RichText
- Click Add Custom Button
Bullet points and numbering icons change colour

shenchauhan replied:
If the toolbar is wider than the parent hosting it, it would be great to have something that indicates there is more.

WilliamABradley replied:
Isn't that what the scrollbar is for? Something like that sounds difficult to implement.

WilliamABradley commented:
@shenchauhan I have fixed the bug with the List Icons, the Link creation, and the Value converter, perhaps the More buttons indicator could be something to be added in a future PR?

shenchauhan commented:
@WilliamABradley  Agreed, approved and merging!! Nice work! 👍

nmetulev commented:
Congrats @WilliamABradley 🥇
